File: app/src/main/java/me/zipi/navitotesla/service/PoiFinder.java
package me.zipi.navitotesla.service;

import java.io.IOException;
import java.util.List;

import me.zipi.navitotesla.exception.DuplicatePoiException;
import me.zipi.navitotesla.model.Poi;

public interface PoiFinder {
    default String findPoiAddress(String poiName) throws DuplicatePoiException, IOException {
        List<Poi> listPoi = listPoiAddress(poiName);
        String address = "";
        int sameCount = 0;
        for (Poi poi : listPoi) {
            if (poi.getPoiName().equalsIgnoreCase(poiName)) {
                sameCount++;
                address = poi.getFinalAddress();
            }
        }
        if (sameCount > 1) {
            // 중복지명 전송 안함
            throw new DuplicatePoiException(poiName);
        }

        return address;
    }

    String parseDestination(String notificationText);

    List<Poi> listPoiAddress(String poiName) throws IOException;

    boolean isIgnore(String notificationTitle, String notificationText);
}

File: app/src/main/kotlin/me/zipi/navitotesla/service/poifinder/PoiFinder.kt
package me.zipi.navitotesla.service.poifinder

import me.zipi.navitotesla.exception.DuplicatePoiException
import me.zipi.navitotesla.model.Poi
import java.io.IOException

interface PoiFinder {
    @Throws(DuplicatePoiException::class, IOException::class)
    suspend fun findPoi(
        poiName: String,
        packageName: String = "",
    ): Poi {
        val poiList = listPoiAddress(poiName).filter { it.poiName.equals(poiName, ignoreCase = true) }

        return if (poiList.size > 1) {
            throw DuplicatePoiException(poiName, poiList.map { it.copy(isDuplicate = true, packageName = packageName) })
        } else if (poiList.size == 1) {
            poiList[0].copy(packageName = packageName)
        } else {
            Poi()
        }
    }

    fun parseDestination(notificationText: String): String

    @Throws(IOException::class)
    suspend fun listPoiAddress(poiName: String): List<Poi>

    /** null 이면 처리, 그 외에는 건너뛴 사유. */
    fun ignoreReason(
        notificationTitle: String,
        notificationText: String,
    ): IgnoreReason?

    fun isIgnore(
        notificationTitle: String,
        notificationText: String,
    ): Boolean = ignoreReason(notificationTitle, notificationText) != null

    /** 접근성으로 모아 둔 목적지를 버린다. */
    fun consumeCapturedDestination() {}
}
